lib/execution_engine2/sdk/job_submission_parameters.py

```diff
@@ -94,5 +106,102 @@ def __hash__(self):
                 self.as_user,
                 self.ignore_concurrency_limits,
                 self.scheduler_requirements,
+                self.debug_mode,
+            )
+        )
+
+
+# move this function somewhere else?
+def _is_valid_UPA(upa: str) -> (str, bool):
+    # returns an empty string if not a valid upa
+    if upa is None or not upa.strip():
+        return "", False
+    parts = [p.strip() for p in upa.split("/")]
+    if not len(parts) == 3:
+        return "", False
+    for p in parts:
+        try:
+            int(p)
+        except ValueError:
+            return "", False
+    return "/".join(parts), True
+
+
+class JobSubmissionParameters:
+    """
+    Parameters for submitting a job to a job scheduler.
+    """
+
+    def __init__(
+        self,
+        job_id: str,
+        app_info: AppInfo,
+        job_reqs: JobRequirements,
+        user_creds: UserCreds,
+        parent_job_id: str = None,
+        wsid: int = None,
+        source_ws_objects: List[str] = None,
+    ):
+        """
+        Create the parameters.
+
+        job_id - the ID of the job.
+        app_info - information about the application to be run.
+        job_reqs - requirements for the job.
+        user_creds - user credentials.
+        parent_job_id - the ID of the parent job to this job, if any.
+        wsid - the ID of the workspace with which the job is associated, if any.
+        source_ws_objects - workspace objects that are part of the job input.
+        """
+        self.job_id = _check_string(job_id, "job_id")
+        self.app_info = _not_falsy(app_info, "app_info")
+        self.job_reqs = _not_falsy(job_reqs, "job_reqs")
+        self.user_creds = _not_falsy(user_creds, "user_creds")
+        self.parent_job_id = _check_string(
+            parent_job_id, "parent_job_id", optional=True
+        )
+        self.wsid = _gt_zero(wsid, "wsid", optional=True)
+        source_ws_objects = source_ws_objects if source_ws_objects else []
+        for i, ref in enumerate(source_ws_objects):
+            upa, is_valid = _is_valid_UPA(ref)
+            if not is_valid:
+                raise IncorrectParamsException(
+                    f"source_ws_objects index {i}, '{ref}', "
+                    + "is not a valid Unique Permanent Address"
+                )
+            source_ws_objects[i] = upa
+        self.source_ws_objects = tuple(source_ws_objects)
+
+    def __eq__(self, other):
+        if type(self) == type(other):
+            return (
+                self.job_id,
+                self.app_info,
+                self.job_reqs,
+                self.user_creds,
+                self.parent_job_id,
+                self.wsid,
+                self.source_ws_objects,
```

Comment:
at some point, that list of properties could be factored out, since it's used in `__hash__` and `__eq__`

(I would already have done it, but I have a fairly low tolerance for repetition)

Reply:
done